internal/controllers/v1beta1/gateway_test.go

```diff
@@ -2,20 +2,435 @@ package v1beta1
+	assert.NoError(t, err)
+	assert.Len(t, certList.Items, 1)
+}
+
+func TestGatewayReconcile_CreateCertificateUsingCredentialName(t *testing.T) {
+	t.Parallel()
+	helper := NewTestHelperWithGateways()
+	assertCreateCertificateCalled(t, helper)
+	cert, err := helper.CertClient.CertmanagerV1().Certificates(TestCertNamespace).Get(context.TODO(), TestCertificateName, metav1.GetOptions{})
+	assert.NoError(t, err)
+	assert.NotNil(t, cert)
+}
+
+func TestGatewayReconcile_CreateCertificateLabeledAsManaged(t *testing.T) {
+	t.Parallel()
+	helper := NewTestHelperWithGateways()
+	assertCreateCertificateCalled(t, helper)
+	cert, err := helper.CertClient.CertmanagerV1().Certificates(TestCertNamespace).Get(context.TODO(), TestCertificateName, metav1.GetOptions{})
+	assert.NoError(t, err)
+	assert.NotNil(t, cert)
+	assert.Equal(t, fmt.Sprintf("%s.%s", TestGatewayName, TestNamespace), cert.Labels[ManagedLabel])
+}
+
+func TestGatewayReconcile_CreateCertificateWithHosts(t *testing.T) {
+	t.Parallel()
+	helper := NewTestHelperWithGateways()
+	assertCreateCertificateCalled(t, helper)
+	cert, err := helper.CertClient.CertmanagerV1().Certificates(TestCertNamespace).Get(context.TODO(), TestCertificateName, metav1.GetOptions{})
+	assert.NoError(t, err)
+	assert.Equal(t, []string{"test1.example.com", "test2.example.com"}, cert.Spec.DNSNames)
+}
+
+func TestGatewayReconcile_CreatCertificateWithDefaultIssuer(t *testing.T) {
+	t.Parallel()
+	helper := NewTestHelperWithGateways()
+	assertCreateCertificateCalled(t, helper)
+	cert, err := helper.CertClient.CertmanagerV1().Certificates(TestCertNamespace).Get(context.TODO(), TestCertificateName, metav1.GetOptions{})
+	assert.NoError(t, err)
+	assert.Equal(t, "default", cert.Spec.IssuerRef.Name)
+	assert.Equal(t, "ClusterIssuer", cert.Spec.IssuerRef.Kind)
+}
+
+func TestGatewayReconcile_CreatCertificateWithClusterIssuerFromGatewayAnnotation(t *testing.T) {
+	t.Parallel()
+	helper := NewTestHelperWithGateways(WithAnnotations(map[string]string{
+		IssuerAnnotation: "testissuer",
+	}))
+	assertCreateCertificateCalled(t, helper)
+	cert, err := helper.CertClient.CertmanagerV1().Certificates(TestCertNamespace).Get(context.TODO(), TestCertificateName, metav1.GetOptions{})
+	assert.NoError(t, err)
+	assert.Equal(t, "testissuer", cert.Spec.IssuerRef.Name)
+	assert.Equal(t, "ClusterIssuer", cert.Spec.IssuerRef.Kind)
+}
+
+func TestGatewayReconcile_SkipCertificateForTLSModeNotSimple(t *testing.T) {
+	t.Parallel()
+	helper := NewTestHelperWithGateways(AppendServer(&networkingv1beta1.Server{
+		Hosts: []string{"pass.example.com"},
+		Tls: &networkingv1beta1.ServerTLSSettings{
+			Mode: networkingv1beta1.ServerTLSSettings_AUTO_PASSTHROUGH,
+		},
+	}))
+	_, err := helper.Controller.Reconcile(context.TODO(), reconcileRequest())
+	assert.NoError(t, err)
+	certList, err := helper.CertClient.CertmanagerV1().Certificates(TestCertNamespace).List(context.TODO(), metav1.ListOptions{})
+	assert.NoError(t, err)
+	assert.Len(t, certList.Items, 1)
```

Comment:
I might be confused, but this should be 0, right?

Perhaps `assertCreateCertificateCalled` can be modified to take in a `int` argument to compare against `helper.Controller.CreateCalled`, and it will be up to the test to pass in how many times it expects Create to be called.

Reply:
There is a default Gateway.  I can update the test description and remove the default host to make this test more clear.